### PR TITLE
Added settings for max volumetric flow rate

### PR DESCRIPTION
**Pull request by mjc-506**
# Description

Added settings for max volumetric flow rate

This fixes https://github.com/Ultimaker/Cura/issues/5248 (together with another PR for CuraEngine)

## Type of change

- [ ] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

A new "Max filament flow rate" setting has been added, defaulting to 0.0 (no effect). Setting this to (say) 10mm3/sec to suit a particular extruder reduces the print head velocity to remain within this limit (especially useful with dymanic layer heights, or changing line widths etc).

Slicing STLs with this PR (and the CuraEngine one) results in slightly different Gcodes with different velocities, respecting the limit set.

**Test Configuration**:
* Operating System: Linux
* Cura 5.30 git
* CuraEngine 5.30 git

# Checklist:
- [X] My code follows the style guidelines of this project as described in [UltiMaker Meta](https://github.com/Ultimaker/Meta) and [Cura QML best practices](https://github.com/Ultimaker/Cura/wiki/QML-Best-Practices)
- [X] I have read the [Contribution guide](https://github.com/Ultimaker/Cura/blob/main/contributing.md) 
- [X] I have commented my code, particularly in hard-to-understand areas
- [NA] I have uploaded any files required to test this change

### CODE REVIEW

**Comment by mashuptwice** [comment]
Thanks for putting in the work! I'll figure out

**Comment by mashuptwice** [comment]
After compiling and having a look at the code, this PR really just adds the setting for max VMS, bot none of the functionality.

Screenshot_20230209_031458
Screenshot_20230209_031436

Running a diff on two coarsely sliced files shows that all the feedrates are exactly the same. 

@mjc-506 Did you even test your work before creating this PR?

File: resources/definitions/fdmprinter.def.json [hunk 1]
```diff
@@ -2528,6 +2528,18 @@
                     "settable_per_extruder": false,
                     "settable_per_meshgroup": false
                 },
+                "material_max_mm3_per_sec":
+                {
+                    "label": "Max Filament Flow Rate",
+                    "description": "When non-zero, sets the maximum allowed filament flow rate (in mm^3/S). The filament flow rate of each printed line is calculated as layer height * line width * print speed. The flow rate limit is enforced by reducing the print speed. Note that these speed reductions are not taken into account when calculating the print time.",
+                    "unit": "mm³/s",
```
Comment (casperlamboo) [inline resources/definitions/fdmprinter.def.json:2535]: You should use the unicode for the cubic $^3$ notation
```suggestion
                    "unit": "mm\u00b3/s",
```

File: resources/definitions/fdmprinter.def.json [hunk 2]
```diff
@@ -2528,6 +2528,18 @@
                     "settable_per_extruder": false,
                     "settable_per_meshgroup": false
                 },
+                "material_max_mm3_per_sec":
+                {
+                    "label": "Max Filament Flow Rate",
+                    "description": "When non-zero, sets the maximum allowed filament flow rate (in mm^3/S). The filament flow rate of each printed line is calculated as layer height * line width * print speed. The flow rate limit is enforced by reducing the print speed. Note that these speed reductions are not taken into account when calculating the print time.",
```
Comment (casperlamboo) [inline resources/definitions/fdmprinter.def.json:2534]: Made the description less verbose, and more consistent with the other setting descriptions
```suggestion
                    "description": "When greater then zero, the maximum allowed flow rate is limited to Max Filament Flow Rate (in mm\u00b3/s). The flow rate limit is enforced by reducing the print speed.",
```

**Review by casperlamboo — CHANGES_REQUESTED** [review]
Contribution looks good! have some minor remarks, if possible please review them yourself as well.

**Comment by casperlamboo** [comment]
See reply at https://github.com/Ultimaker/CuraEngine/pull/1814#issuecomment-1584200537

>  Hi @mjc-506 sorry for taking this long to get back to you. We have had quite some internal discussions about this feature, and unfortunately we have decided to take a different approach.
>
> We are afraid that, when using this new feature, the print behavior will become unpredictable. Instead we want to introduce a new feature that gives the user more insight whether flow jumps are violated, and can consequently change settings to make sure flow stays within an acceptable range.
>
> Hope you understand our decision.

**Pull request closed without merging**